ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -31,215 +81,264 @@ where
             g: g.into_iter().collect(),
         }
     }
+
+    /// Distributed Zero Knowledge Proofs algorithm drawn from
+    /// `https://eprint.iacr.org/2023/909.pdf`
+    ///
+    /// Uses Andy's borrow trick to work with references and owned values :)
+    pub fn compute_proof<J, B>(
```

Comment:
ProofGenerator was a wrapper around a vec and would generate the proof from the vec. This is not what we need. We need to generate at least the first proof from an iterator since we don't want to store a giant vector of field elements. We want to generate the proofs by lazily generating the field elements.

First I changed ProofGenerator to use an IntoIter. But that also did not make much sense since there were a lot of generics involved that did not make sense when generating the first proof. I think the structure now makes a lot of sense. If you want to have a zero knowledge proof, you call compute_proof on the zero-knowledge proof struct. The default method gives a an all 0 zero knowledge proof which will also verify for all zero field elements, i.e. default field elements.